cmd/juju/machine/remove.go

```diff
@@ -111,17 +124,27 @@ func (a removeMachineAdapter) destroyMachines(f func(...string) error, machines
 	return results, nil
 }
 
-func (c *removeCommand) getRemoveMachineAPI() (RemoveMachineAPI, error) {
-	if c.api != nil {
-		return c.api, nil
+func (c *removeCommand) getAPIRoot() (api.Connection, error) {
+	if c.apiRoot != nil {
+		return c.apiRoot, nil
 	}
-	root, err := c.NewAPIRoot()
+	return c.NewAPIRoot()
+}
+
+func (c *removeCommand) getRemoveMachineAPI() (RemoveMachineAPI, error) {
+	root, err := c.getAPIRoot()
 	if err != nil {
 		return nil, err
 	}
-	if root.BestFacadeVersion("MachineManager") >= 3 {
+	if root.BestFacadeVersion("MachineManager") < 4 && c.KeepInstance {
+		return nil, errors.New("this version of Juju doesn't support --keep-instance")
```

Comment:
Why have this knowledge in two places? You already return an error from the client facade.